### PR TITLE
Add bytecode profiler

### PR DESCRIPTION
**Pull request by jmerle**
This PR adds a bytecode profiler to the engine and a visualizer of the recorded data to the client. The visualizer uses [speedscope](https://github.com/jlfwong/speedscope) to display the results. Here's a little demo: [https://www.youtube.com/watch?v=j5noBI3exaQ](https://www.youtube.com/watch?v=j5noBI3exaQ).

This PR is accompanied by https://github.com/battlecode/battlecode20-scaffold/pull/9 because the profiler is toggled by a Gradle property. This is also written out on the Help tab of the sidebar.

The bytecode profiling _does not_ affect actual bytecode usage, the entire profiling part is wrapped around the bytecode counting so the profiling isn't seen as player code.

Recorded data is stored in the bc20 file, for which I extended the MatchFooter table in the FlatBuffers schema. bc20 files with profiling data can grow somewhat big (I've seen 10+ mb playing my sprint bot against itself), but since it's not supposed to be ran in scrimmages/tournaments that shouldn't be a problem. What also helps is that the profiler only profiles player methods, so if RobotController.run() calls a method on (for example) an ArrayList, the call to ArrayList is not monitored and the bytecode used is simply added to the record of the RobotController.run method.

### CODE REVIEW

File: client/visualizer/electron-main.js [hunk 1]
```diff
@@ -52,6 +52,11 @@ function createWindow () {
   });
 }
 
+// This is needed so that the speedscope iframe on the Profiler tab does not throw cross-origin errors
+// The speedscope iframe loads speedscope using the file:// protocol which considers all URLs to have separate origins
+// The only way to get around this is to tell Chrome to allow it anyways
+electron.app.commandLine.appendSwitch('disable-site-isolation-trials');
```
Comment (arvid220u) [inline client/visualizer/electron-main.js]: this seems less than ideal but maybe it is inevitable

(just commenting so i remember to go back to this)

File: client/visualizer/README.md [hunk 2]
```diff
@@ -54,7 +54,7 @@ $ npm run prod
 This will bundle up all of the assets you want in the `bc20` folder. You can then embed the client in any web page you want:
 
 - Mount the assets in `bc20` at `/bc20/` on your webserver.
-- `<script src="/bc20/bundle.js"></script>`
+- `<script src="/bc20/app.js"></script>`
```
Comment (arvid220u) [inline client/visualizer/README.md]: check this

File: client/visualizer/index.html [hunk 3]
```diff
@@ -6,7 +6,7 @@
     </head>
     <body>
         <div id='client'/>
-        <script type="text/javascript" src="bc20/bundle.js" charset="utf-8"></script>
+        <script type="text/javascript" src="bc20/app.js" charset="utf-8"></script>
```
Comment (arvid220u) [inline client/visualizer/index.html]: bundle.js app.js change

File: client/visualizer/package.json [hunk 4]
```diff
@@ -57,7 +58,7 @@
     "productName": "Battlecode Client",
     "compression": "maximum",
     "files": [
-      "bc20/*",
+      "bc20/**/*",
```
Comment (arvid220u) [inline client/visualizer/package.json]: hmm

File: client/visualizer/src/game/gamearea/gamearea.ts [hunk 5]
```diff
@@ -125,17 +127,22 @@ export default class GameArea {
...
+        default:
+          this.wrapper.appendChild(this.canvas); // TODO: Only append if a game is available in client.games
+          console.log("Now a game");
       }
     }
     
     if(shouldListen) {
       window.onresize = function() {
-        var wrapper = <HTMLDivElement> document.getElementById("canvas-wrapper")
+        var wrapper = <HTMLDivElement> document.getElementById("canvas-wrapper");
```
Comment (arvid220u) [inline client/visualizer/src/game/gamearea/gamearea.ts]: :0

File: client/visualizer/src/game/sidebar/matchrunner.ts [hunk 6]
```diff
@@ -38,10 +38,11 @@ export default class MatchRunner {
   // Match information
   private teamA: HTMLSelectElement;
   private teamB: HTMLSelectElement;
+  private profilerEnabled: HTMLInputElement;
   private selectAllMaps: HTMLButtonElement;
   private deselectAllMaps: HTMLButtonElement;
   private runMatch: HTMLButtonElement;
-  private refreshButton: HTMLButtonElement
+  private refreshButton: HTMLButtonElement;
```
Comment (arvid220u) [inline client/visualizer/src/game/sidebar/matchrunner.ts]: :0

File: client/visualizer/src/main/sidebar.ts [hunk 7]
```diff
@@ -125,8 +129,13 @@ export default class Sidebar {
     <b class="blue">Keyboard Shortcuts</b><br>
     LEFT - Step Back One Turn<br>
     RIGHT - Step Forward One Turn<br>
+<<<<<<< HEAD
```
Comment (arvid220u) [inline client/visualizer/src/main/sidebar.ts]: let's change this

File: client/visualizer/src/main/sidebar.ts [hunk 8]
```diff
@@ -301,8 +319,13 @@ export default class Sidebar {
       case Mode.QUEUE:
         this.innerDiv.appendChild(this.matchqueue.div);
         break;
+      case Mode.PROFILER:
+        this.innerDiv.append(this.profiler.div);
+        break;
     }
 
-    // this.cb();
+    if (this.cb !== undefined) {
```
Comment (arvid220u) [inline client/visualizer/src/main/sidebar.ts]: check this

File: client/visualizer/src/static/css/style.css [hunk 9]
```diff
@@ -86,11 +101,17 @@ form {
 
 #gamearea canvas {
     max-height: calc(100vh - 75px);
-    max-width: calc(100vw - 320px);
+    max-width: calc(100vw - 345px);
```
Comment (arvid220u) [inline client/visualizer/src/static/css/style.css]: .

File: client/visualizer/webpack.config.js [hunk 10]
```diff
@@ -10,11 +10,12 @@ var conf = {
   context: path.resolve(__dirname, 'src'),
   entry: {
     app: './app.ts',
+    profiler: './profiler.ts'
   },
   output: {
     path: path.resolve(__dirname, 'bc20'),
     publicPath: '/bc20/',
-    filename: 'bundle.js'
+    filename: '[name].js'
```
Comment (arvid220u) [inline client/visualizer/webpack.config.js]: ohhhh i see

File: client/visualizer/webpack.config.js [hunk 11]
```diff
@@ -27,6 +28,26 @@ var conf = {
...
+
+          return content;
+        }
+      }
+    ])
+  ],
+  devServer: {
+    // Required to ensure the files copied by the CopyWebpackPlugin are copied when running the dev server
+    writeToDisk: filePath => filePath.includes('speedscope/')
+  },
```
Comment (arvid220u) [inline client/visualizer/webpack.config.js]: hmm why is this necessary

File: engine/src/main/battlecode/instrumenter/bytecode/InstrumentingMethodVisitor.java [hunk 12]
```diff
@@ -153,6 +152,9 @@ public void visitMaxs(int maxStack, int maxLocals) {
 
         boolean anyTryCatch = tryCatchBlocks.size() > 0;
 
+        // must be called before addDebugHandler() so debug methods are profiled properly
+        addEnterMethodHandler();
```
Comment (arvid220u) [inline engine/src/main/battlecode/instrumenter/bytecode/InstrumentingMethodVisitor.java]: so i guess even if instrumenting is disabled we inject the instrumenter code?

(again this is only my commments so i can figure out what happens here)

File: engine/src/main/battlecode/instrumenter/bytecode/InstrumentingMethodVisitor.java [hunk 13]
```diff
@@ -185,6 +187,36 @@ private static AbstractInsnNode nextInstruction(AbstractInsnNode n) {
         return n;
     }
 
+    private void addEnterMethodHandler() {
```
Comment (arvid220u) [inline engine/src/main/battlecode/instrumenter/bytecode/InstrumentingMethodVisitor.java]: for later

File: engine/src/main/battlecode/instrumenter/inject/RobotMonitor.java [hunk 14]
```diff
@@ -129,11 +138,16 @@ public static void incrementBytecodes(int numBytecodes) {
                 bytecodesLeft = Integer.MIN_VALUE;
             }
 
+            // the profiler isn't enabled in ranked matches, so the integer overflow isn't really a problem
+            if (profiler != null) {
+                profiler.incrementBytecodes(numBytecodes + bytecodesToRemove);
```
Comment (arvid220u) [inline engine/src/main/battlecode/instrumenter/inject/RobotMonitor.java]: we probably still want to avoid the overflow

File: engine/src/main/battlecode/instrumenter/inject/RobotMonitor.java [hunk 15]
```diff
@@ -235,6 +249,32 @@ public static long getRandomSeed() {
...
     }
 
+    /**
+     * Called at the start of a method. Used by the profiler.
+     *
+     * THIS METHOD IS CALLED BY THE INSTRUMENTER.
+     *
+     * @param methodName the name of the method that is being entered
+     */
+    public static void enterMethod(String methodName) {
```
Comment (arvid220u) [inline engine/src/main/battlecode/instrumenter/inject/RobotMonitor.java]: oh ok this seems fine

**Comment by arvid220u** [comment]
sorry for the many comments, just want to read through everything. i really like this change though and we'll probably merge it tomorrow!

File: client/visualizer/electron-main.js (hunk 1 above)
Comment (jmerle) [reply client/visualizer/electron-main.js]: When this line is removed, the Profiler tab works when the application is first started but breaks when the user hits Ctrl+R. There may be other solutions to this, but it already took me 30 minutes to fix it using this line and I also had a sprint bot to work on :)

File: client/visualizer/package.json (hunk 4 above)
Comment (jmerle) [reply client/visualizer/package.json]: The CopyWebpackPlugin now copies the speedscope code to "bc20/speedscope". "bc20/*" non-recursively takes all contents inside the "bc20/" directory and skips the files inside "bc20/speedscope/" because that's a different directory. I might be wrong here though.

File: client/visualizer/src/game/gamearea/gamearea.ts (hunk 5 above)
Comment (jmerle) [reply client/visualizer/src/game/gamearea/gamearea.ts]: I use IntelliJ and it highlighted all the missing semicolons, so I added some as I read over them and eventually gave up because they are used wayyyy too inconsistent :P

File: client/visualizer/src/main/sidebar.ts (hunk 7 above)
Comment (jmerle) [reply client/visualizer/src/main/sidebar.ts]: Whoops, forgot that merge conflict, can fix in ~2 hours

File: client/visualizer/src/static/css/style.css (hunk 9 above)
Comment (jmerle) [reply client/visualizer/src/static/css/style.css]: Sidebar has been made bigger to make space for the Profiler tab. As a consequence, the gamearea canvas has been made smaller so it doesn't flow underneath the sidebar.

File: client/visualizer/webpack.config.js (hunk 10 above)
Comment (jmerle) [reply client/visualizer/webpack.config.js]: Sidebar has been made bigger to make space for the Profiler tab. As a consequence the gamearea canvas has been made smaller.

**Comment by jmerle** [comment]
That's what comments are for :)

I've replied to some of them to make it more clear and will fix the merge conflicts in about ~2 hours.

File: client/visualizer/webpack.config.js (hunk 11 above)
Comment (jmerle) [reply client/visualizer/webpack.config.js]: When using webpack-dev-server (with `npm run watch` in the client directory) the CopyWebpackPlugin doesn't actually copy the speedscope files without this line because webpack-dev-server does everything in-memory. This causes problems because those speedscope files are needed by the iframe visible when the Profiler tab is active. By using `writeToDisk` we can tell webpack-dev-server to write certain files to disk even when running primarily in-memory (in this case all the speedscope files).

File: engine/src/main/battlecode/instrumenter/bytecode/InstrumentingMethodVisitor.java (hunk 12 above)
Comment (jmerle) [reply engine/src/main/battlecode/instrumenter/bytecode/InstrumentingMethodVisitor.java]: Not anymore

**Comment by jmerle** [comment]
I've fixed the merge conflicts, fixed the potential integer overflow and made it so that it doesn't instrument the player code with profiling related code when profiling is disabled.

File: client/visualizer/electron-main.js (hunk 1 above)
Comment (arvid220u) [reply client/visualizer/electron-main.js]: oh i see, let's just leave this here then

File: client/visualizer/webpack.config.js (hunk 11 above)
Comment (arvid220u) [reply client/visualizer/webpack.config.js]: i see i see

File: client/visualizer/src/main/sidebar.ts (hunk 8 above)
Comment (jmerle) [reply client/visualizer/src/main/sidebar.ts]: this.cb() is set to update the gamearea content based on the currently open tab. I guess this was commented out after removing the map editor (currently the replay viewer is always shown) and I added it back in in order to show the speedscope iframe when the Profiler tab is active. Not sure about the `!== undefined` check, but I'm guessing it caused an error without it.

Comment (arvid220u) [reply client/visualizer/src/main/sidebar.ts]: Makes sense, thanks

**Comment by arvid220u** [comment]
@jmerle Have you also been getting `Exception in thread "main" java.lang.OutOfMemoryError: Java heap space` when running games with the profiler? I'm running a game with a lot of robots though so maybe that is why

**Comment by arvid220u** [comment]
Amazing work! Thank you so much for doing this. I've merged this and will push it out tonight.
